core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -167,8 +177,12 @@ protected void validate(TableMetadata currentMetadata) {
   @Override
   public Snapshot apply() {
     refresh();
-    Long parentSnapshotId = base.currentSnapshot() != null ?
-        base.currentSnapshot().snapshotId() : null;
+    Long parentSnapshotId = null;
+    if(toBranch != null){
+      parentSnapshotId = base.ref(toBranch).snapshotId();
+    } else {
+      parentSnapshotId = base.currentSnapshot() != null ? base.currentSnapshot().snapshotId() : null;
+    }
```

Comment:
We can default assign the parentSnapshotId. I also think the SnapshotProducer can maintain the current branch itself as a field, (defaulting to main and then getting overrode in the toBranch).